### PR TITLE
Register futures in async registry

### PR DESCRIPTION
**Pull request by jvolmer**
With this PR, the async registry includes not only async coroutines (which are at this point still very rarely used in our codebase), but also futures.
Futures that are directly resolved are not added to the registry.
The source location is handed through, such that the registry gets the location where the future object was initially created instead of a location inside the futures library.

### CODE REVIEW

**Review by jbajic — APPROVED** [review]

File: lib/Basics/FutureSharedLock.h [hunk 1]
```diff
@@ -180,8 +180,7 @@ struct FutureSharedLock {
 
     void unlock() {
       std::lock_guard lock(_mutex);
-      TRI_ASSERT(_lockCount > 0);
-      if (--_lockCount == 0 && !_queue.empty()) {
+      if (_lockCount > 0 && --_lockCount == 0 && !_queue.empty()) {
```
Comment (maierlars) [inline lib/Basics/FutureSharedLock.h:183]: Why was this changed? I'd rather make the `TRI_ASSERT` into a `ADB_PROD_ASSERT`

**Review by maierlars — APPROVED** [review]
